app/models/fact_value.rb

```diff
@@ -81,13 +81,19 @@ def self.count_each(fact, options = {})
     output
   end
 
-  def self.build_facts_hash(facts)
-    hosts = Host.where(:id => facts.group_by(&:host_id).keys).all
-
+  def self.build_facts_hash(facts, host_id = nil)
     hash = {}
+    if host_id
+      hosts = Host.where(:id => host_id).empty? ? Host.where(:name => host_id) : Host.where(:id => host_id)
+    else
+      hosts =  Host.where(:id =>facts.map(&:host_id).uniq)
```

Comment:
this will load all facts to memory just to calculate the ids, you can do this in the db using `facts.distinct.pluck(:host_id)`.